Security_Engine/Compute/CameraFieldOfView.cs

```diff
@@ -0,0 +1,351 @@
+
+        private static Polyline FixAndProject(this Polyline obstacle, Plane cameraPlane, double tolerance)
+        {
+            if (obstacle.IsInPlane(cameraPlane))
+                return obstacle;
+
+            if (!obstacle.IsClosed(tolerance) || !obstacle.IsPlanar(tolerance))
+            {
+                Base.Compute.RecordWarning("Obstacle polyline is open and not planar. It will be closed and projected on camera plane.");
+                List<Point> projPoints = obstacle.ControlPoints.Select(x => x.Project(cameraPlane)).ToList();
+                return BH.Engine.Geometry.Create.Polyline(projPoints);
+            }
+            else if (obstacle.Normal().IsEqual(cameraPlane.Normal, tolerance) || obstacle.Normal().IsEqual(-cameraPlane.Normal, tolerance))
+            {
+                Base.Compute.RecordWarning("Obstacle polyline is parallel, but not in camera plane. It will be projected on camera plane.");
+                return obstacle.Project(cameraPlane);
+            }
+            else
+            {
+                Base.Compute.RecordWarning("Polyline is in different plane than camera plane. It will be projected on camera plane.");
+                return obstacle.Project(cameraPlane);
+            }
+        }
+
+        /***************************************************/
+
+        [Description("Split line in given obstacles.")]
+        [Input("rayLine", "Line to split.")]
+        [Input("obstacles", "Polyline objects to split the line in.")]
+        [Output("splitLines", "Lines splited by the obstacles.")]
+        private static List<Line> SplitLinesInObstacles(this Line line, List<Polyline> obstacles)
+        {
+            List<Point> splitPoints = new List<Point>();
+            foreach (Polyline obstacle in obstacles)
+            {
+                foreach (Line obstLine in obstacle.SubParts())
+                {
+                    Point point = line.LineIntersection(obstLine);
+                    if (point != null)
+                        splitPoints.Add(point);
+                }
+            }
+            return line.SplitAtPoints(splitPoints);
+        }
+
+        /***************************************************/
+
+        [Description("Check if line is inside obstacle.")]
+        [Input("line", "Line to check if it's inside obstacle.")]
+        [Input("obstacles", "Polyline objects to check if line is inside.")]
+        [Output("bool", "Boolean result of the checking.")]
+        private static bool IsInsideObstacle(this Line line, List<Polyline> obstacles)
+        {
+            foreach (Polyline obstacle in obstacles)
+            {
+                foreach (Line obstLine in obstacle.SubParts())
+                {
+                    if (obstLine.IsOnCurve(line.Start) && obstLine.IsOnCurve(line.End))
+                        return false;
+                }
+                if (obstacle.IsContaining(new List<Point> { line.Centroid() }, true))
```

Comment:
It is worth doing some profiling in case the method is slow: I am pretty sure `IsContaining` will consume a big parf of the entire execution time, which could be addressed by the tweaks considered (implemented?) by @IsakNaslundBh on the other occasion.

Reply:
Moved to issue #2890.